wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,866 @@
+# cuML Python Estimators Developer Guide
+
+This guide is meant to help developers follow the correct patterns when creating/modifying any cuML Estimator object and ensure a uniform cuML API.
+
+**Note:** This guide is long, because it includes internal details on how cuML manages input and output types for advanced use cases. But for the vast majority of estimators, the requirements are very simple and can follow the example patterns shown below in the [Quick Start Guide](#quick-start-guide).
+
+Start by reading the following:
+
+1. [SkLearn Estimator Docs](https://scikit-learn.org/stable/developers/develop.html)
+   1. cuML Estimator design follows Sklearn very closely. We will only cover portions where our design differs from this document
+   2. Pay close attention to these sections as these topics have caused pain points in the past:
+      1. [Instantiation](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+      2. [Estimated Attributes](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+      3. [`get_params` and `set_params`](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+      4. [cloning](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
+2. [SkLearn Docstring Guide](https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation)
+   1. We follow the same guidlines for speficying array-like objects, array shapes, dtypes, and default values
+
+## Quick Start Guide
+
+At a high level, all cuML Estimators must:
+1. Inherit from `cuml.common.base.Base`
+   ```python
+   from cuml.common.base import Base
+
+   class MyEstimator(Base):
+      ...
+   ```
+2. Follow the Scikit-learn estimator guidelines found [here](https://scikit-learn.org/stable/developers/develop.html)
+3. Include the `Base.__init__()` arguments available in the new Estimator's `__init__()` 
+   ```python
+   class MyEstimator(Base):
+
+      def __init__(self, *, extra_arg=True, handle=None, verbose=False, output_type=None):
+         super().__init__(handle=handle, verbose=verbose, output_type=output_type)
+         ...
+   ```
+4. Declare each array-like attribute the new Estimator will compute as a class variable for automatic array type conversion
+   ```python
+   from cuml.common.array_descriptor import CumlArrayDescriptor
+
+   class MyEstimator(Base):
+
+      labels_ = CumlArrayDescriptor()
+
+      def __init__(self):
+         ...
+   ```
+5. Add input and return type annotations to public API functions OR wrap those functions explicitly with conversion decorators (see [Estimator Methods](#estimator-methods) for non-standard use cases)
```

Comment:
Perhaps a small example here of a non-standard use case?

Reply:
I was trying to keep the quick start guide short so instead of adding in a non-standard use case, I linked to one later in the document. Do you feel strongly that it should be included here?

Reply:
Changed it to a more specific section header.